storage/pebble/registers_pruner_test.go

```diff
@@ -0,0 +1,397 @@
+}
+
+// cleanupPruner stops the pruner and verifies there are no errors in the error channel.
+func cleanupPruner(t *testing.T, pruner *RegisterPruner, cancel context.CancelFunc, errChan <-chan error) {
+	cancel()
+	<-pruner.Done()
+
+	select {
+	case err := <-errChan:
+		require.NoError(t, err)
+	default:
+	}
+}
+
+// straightPruneTestCase initializes and returns a testCase with predefined data for straight pruning.
+func straightPruneTestCase() testCase {
+	initialData := emptyRegistersData(12)
+
+	key1 := flow.RegisterID{Owner: "owner1", Key: "key1"}
+	key2 := flow.RegisterID{Owner: "owner2", Key: "key2"}
+	key3 := flow.RegisterID{Owner: "owner3", Key: "key3"}
+
+	value1 := []byte("value1")
+	value2 := []byte("value2")
+	value3 := []byte("value3")
+
+	// Set up initial register entries for different heights.
+	initialData[3] = flow.RegisterEntries{
+		{Key: key1, Value: value1},
+		{Key: key2, Value: value2},
+	}
+	initialData[4] = flow.RegisterEntries{
+		{Key: key2, Value: value2},
+		{Key: key3, Value: value3},
+	}
+	initialData[7] = flow.RegisterEntries{
+		{Key: key1, Value: value1},
+		{Key: key2, Value: value2},
+		{Key: key3, Value: value3},
+	}
+	initialData[8] = flow.RegisterEntries{
+		{Key: key1, Value: value1},
+		{Key: key3, Value: value3},
+	}
+	initialData[9] = flow.RegisterEntries{
+		{Key: key2, Value: value2},
+		{Key: key3, Value: value3},
+	}
+	initialData[10] = flow.RegisterEntries{
+		{Key: key1, Value: value1},
+		{Key: key2, Value: value2},
+	}
+	initialData[12] = flow.RegisterEntries{
+		{Key: key1, Value: value1},
+		{Key: key3, Value: value3},
+	}
+
+	// Define the expected data after pruning.
+	expectedData := map[uint64]flow.RegisterEntries{
+		7: {
+			{Key: key1, Value: value1}, // keep, first row <= 7
+			{Key: key2, Value: value2}, // keep, first row <= 7
+			{Key: key3, Value: value3}, // keep, first row <= 7
+		},
+		8: {
+			{Key: key1, Value: value1}, // keep, height > 7
+			{Key: key3, Value: value3}, // keep, height > 7
+		},
+		9: {
+			{Key: key2, Value: value2}, // keep, height > 7
+			{Key: key3, Value: value3}, // keep, height > 7
+		},
+		10: {
+			{Key: key1, Value: value1}, // keep, height > 7
+			{Key: key2, Value: value2}, // keep, height > 7
+		},
+		12: {
+			{Key: key1, Value: value1}, // keep, height > 7
+			{Key: key3, Value: value3}, // keep, height > 7
+		},
+	}
+
+	// Define the data that should be pruned (i.e., removed) from the database after pruning.
+	prunedData := map[uint64]flow.RegisterEntries{
+		3: {
+			{Key: key1, Value: value1},
+			{Key: key2, Value: value2},
+		},
+		4: {
+			{Key: key2, Value: value2},
+			{Key: key3, Value: value3},
+		},
+	}
+
+	return testCase{
+		initialData:         initialData,
+		expectedFirstHeight: 7,
+		expectedData:        expectedData,
+		prunedData:          prunedData,
+	}
+}
+
+// testCaseWithDiffHeights initializes and returns a testCase with predefined data for different entries to keep
+func testCaseWithDiffHeights() testCase {
```

Comment:
In addition to these test cases, it's better to add some property based random tests.

One property I can think of is that, pruning would not affect the value at any height above the pruned height.

To implement this, we can add register values at random heights in a range, and then prune by a certain height within that range, and assert that the register value at each height is unchanged. And we can also extend the test case to multiple registers.

Another property is that if the pruning is interrupted at any point, the register values at any height should not be changed.